src/algebra/order/hom/basic.lean

```diff
@@ -85,3 +107,147 @@ by simpa only [div_mul_div_cancel'] using map_mul_le_mul f (a / b) (b / c)
+
+end group_seminorm_class
+
+example [ordered_add_comm_group β] : ordered_add_comm_monoid β := infer_instance
+
+@[to_additive] lemma abs_sub_map_le_div [group α] [linear_ordered_add_comm_group β]
+  [group_seminorm_class F α β] (f : F) (x y : α) : |f x - f y| ≤ f (x / y) :=
+begin
+  rw [abs_sub_le_iff, sub_le_iff_le_add', sub_le_iff_le_add'],
+  exact ⟨le_map_add_map_div _ _ _, le_map_add_map_div' _ _ _⟩
+end
+
+@[to_additive, priority 100] -- See note [lower instance priority]
+instance group_seminorm_class.to_nonneg_hom_class [group α] [linear_ordered_add_comm_monoid β]
+  [group_seminorm_class F α β] :
+  nonneg_hom_class F α β :=
+{ map_nonneg := λ f a, (nsmul_nonneg_iff two_ne_zero).1 $
+    by { rw [two_nsmul, ←map_one_eq_zero f, ←div_self' a], exact map_div_le_add _ _ _ },
+  ..‹group_seminorm_class F α β› }
+
+section group_norm_class
+variables [group α] [ordered_add_comm_monoid β] [group_norm_class F α β] (f : F) {x : α}
+include α β
+
+@[simp, to_additive] lemma map_eq_zero_iff_eq_one : f x = 0 ↔ x = 1 :=
+⟨eq_one_of_map_eq_zero _, by { rintro rfl, exact map_one_eq_zero _ }⟩
+
+@[to_additive] lemma map_ne_zero_iff_ne_one : f x ≠ 0 ↔ x ≠ 1 := (map_eq_zero_iff_eq_one _).not
+
+end group_norm_class
+
+@[to_additive] lemma map_pos_of_ne_one [group α] [linear_ordered_add_comm_monoid β]
+  [group_norm_class F α β] (f : F) {x : α} (hx : x ≠ 1) : 0 < f x :=
+(map_nonneg _ _).lt_of_ne $ ((map_ne_zero_iff_ne_one _).2 hx).symm
+
+/-! ### Ring (semi)norms -/
+
+/-- `ring_seminorm_class F α` states that `F` is a type of `β`-valued seminorms on the ring `α`.
+
+You should extend this class when you extend `ring_seminorm`. -/
+class ring_seminorm_class (F : Type*) (α β : out_param $ Type*) [non_unital_non_assoc_ring α]
+  [ordered_semiring β] extends add_group_seminorm_class F α β, submultiplicative_hom_class F α β
+
+/-- `ring_norm_class F α` states that `F` is a type of `β`-valued norms on the ring `α`.
+
+You should extend this class when you extend `ring_norm`. -/
+class ring_norm_class (F : Type*) (α β : out_param $ Type*) [non_unital_non_assoc_ring α]
+  [ordered_semiring β] extends ring_seminorm_class F α β, add_group_norm_class F α β
+
+/-- `mul_ring_seminorm_class F α` states that `F` is a type of `β`-valued multiplicative seminorms
+on the ring `α`.
+
+You should extend this class when you extend `mul_ring_seminorm`. -/
+class mul_ring_seminorm_class (F : Type*) (α β : out_param $ Type*) [non_assoc_ring α]
+  [ordered_semiring β] extends add_group_seminorm_class F α β, monoid_with_zero_hom_class F α β
+
+/-- `mul_ring_norm_class F α` states that `F` is a type of `β`-valued multiplicative norms on the
+ring `α`.
+
+You should extend this class when you extend `mul_ring_norm`. -/
+class mul_ring_norm_class (F : Type*) (α β : out_param $ Type*) [non_assoc_ring α]
+  [ordered_semiring β] extends mul_ring_seminorm_class F α β, add_group_norm_class F α β
+
+-- TODO: Somehow this does not get inferred.
```

Comment:
Could you explain what this means specifically? Do you have an example of this instance not applying when it should?

Reply:
Aha, this is because of diamond inheritance in instance parameters that depend on an `out_param`: `add_group_seminorm_class.to_nonneg_hom_class` requires a `linear_ordered_add_comm_monoid` instance, and `ring_seminorm_class.to_add_group_seminorm_class` requires a `ordered_semiring` instance. These have a common ancestor in `ordered_add_comm_monoid` so normally what happens is Lean wants to unify `@linear_ordered_add_comm_monoid.to_ordered_add_comm_monoid β ?inst1 =?= @ordered_semiring.to_ordered_add_comm_monoid β ?inst2`, sees that it can unstuck by inferring instances for `?inst1` and `?inst2`, unfolds those instances, compares equality field-wise, and succeeds.

Here, because `β` is an `out_param` of `nonneg_hom_class`, `β` becomes a metavariable that isn't solved until the `ring_seminorm_class` parameter is filled in. But we can't fill that in until we reach `ring_seminorm_class` from `add_group_seminorm_class`. And we can't check that these are defeq because it requires getting unstuck, and it can't unstuck because the instance is full of metavariables, so Lean gives up the equality check.

Hacking on mathlib4#888 has taught me the Lean 4 synthesis algorithm is even worse at instance params depending on `outParam`s so I expect the conclusion would have to be that this instance is required to help Lean get unstuck.

Reply:
Oh dear 😦

Then I guess this PR is ready

Reply:
Let's add a comment:
```suggestion
-- This instance is simply the composition of the instances
-- `add_group_seminorm_class.to_nonneg_hom_class` and
-- `ring_seminorm_class.to_add_group_seminorm_class`. However, it has to be
-- declared explicitly due to hitting an edge case in the inference algorithm:
-- those instances both depend on a different subclass of `ordered_add_monoid β`, but `β` is an
-- `out_param` that cannot be inferred until both instances are applied.
-- So we have to do it manually until Lean becomes smarter.
```

Reply:
I've edited the suggestion to add a pointer to a relevant Zulip thread. If we really want to get this working, I think the right solution is to forbid extending a class in conditions where this limitation applies, and instead use unbundled inheritance.

If you want to add a library note, I'd add it as a separate comment, specifying exactly what goes wrong and some options for dealing with it:
```quote
Diamond inheritance cannot depend on `out_param`s in the following circumstances:
 * there are three classes `top`, `middle`, `bottom`
 * all of these classes have a parameter `(α : out_param _)`
 * all of these classes have an instance parameter `[root α]` that depends on this `out_param`
 * the `root` class has two child classes: `left` and `right`, these are siblings in the hierarchy
 * the instance `bottom.to_middle` takes a `[left α]` parameter
 * the instance `middle.to_top` takes a `[right α]` parameter
 * there is a `leaf` class that inherits from both `left` and `right`.
In that case, given instances `bottom α` and `leaf α`, Lean cannot synthesize a `top α` instance, even though the hypotheses of the instances `bottom.to_middle` and `middle.to_top` are satisfied.

There are two workarounds:
 * You could replace the bundled inheritance implemented by the instance `middle.to_top` with unbundled inheritance implemented by adding a `[top α]` parameter to the `middle` class. This is the preferred option since it is also more compatible with Lean 4, at the cost of being more work to implement and more verbose to use.
 * You could weaken the `bottom.to_middle` instance by making it depend on a subclass of `middle.to_top`'s parameter, in this example replacing `[left α]` with `[leaf α]`.
```